concept/thing/thing_manager/validation/validation.rs

```diff
@@ -0,0 +1,75 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use encoding::value::{label::Label, ValueEncodable};
+use lending_iterator::LendingIterator;
+use storage::snapshot::ReadableSnapshot;
+
+use crate::{
+    error::ConceptReadError,
+    thing::{
+        object::ObjectAPI,
+        thing_manager::{validation::DataValidationError, ThingManager},
+    },
+    type_::{
+        annotation::{AnnotationCategory, AnnotationKey, AnnotationUnique},
+        owns::{Owns, OwnsAnnotation},
+        type_manager::TypeManager,
+        Capability, TypeAPI,
+    },
+};
+
+pub(crate) fn get_label_or_concept_read_err<'a>(
+    snapshot: &impl ReadableSnapshot,
+    type_manager: &TypeManager,
+    type_: impl TypeAPI<'a>,
+) -> Result<Label<'static>, ConceptReadError> {
+    type_.get_label(snapshot, type_manager).map(|label| label.clone())
+}
+
+pub(crate) fn get_label_or_data_err<'a>(
+    snapshot: &impl ReadableSnapshot,
+    type_manager: &TypeManager,
+    type_: impl TypeAPI<'a>,
+) -> Result<Label<'static>, DataValidationError> {
+    get_label_or_concept_read_err(snapshot, type_manager, type_).map_err(DataValidationError::ConceptRead)
+}
+
+pub(crate) fn get_uniqueness_source(
```

Comment:
I didn't do it because:
1. I didn't want to expose it as it should be needed only for the inner logic of us
2. It kinda feels lonely and I have a desire to create multiple `get_***_source` for other annotations (and their final values), like the last `values` narrowing, but it's not needed. 

Feels better in terms of domain, so I'll move this function to `Owns`.

Reply:
Nah, I still want to panic, because `is_unique == true` && `get_uniqueness_source() == None` means broken logic!